### ISSUE TITLE
Not getting same results when using OpenMP Interface and without it.

### ISSUE THREAD

**Issue by fahadzafar** [openMVG/openMVG#878]
So I am testing the library on Mac OS X and Ubuntu.

My Mac OS S didnt have openmp and cmake configured the build without it. Great, works fine.
Then i installed openmp on Ubuntu and ran cmake. Cant turn off the OpenMP flags. Seems like a bug in the cmake script. using 
cmake -dUSE_OPENMP=OFF      or
cmake -dOpenMVG_USE_OPENMP=OFF

does not write a configured script. Here is the cmake error,
ERROR: Unknown CMake command "UPDATE_CACHE_VARIABLE"

// ------------------
Using the same dataset, it creates a mesh in the Mac (without openMP) but not the Ubuntu (with openMP) 

I track every stage and it starts changing here, 

 - PUTATIVE MATCHES - 
Use: exhaustive pairwise matching
Using FAST_CASCADE_HASHING_L2 matcher
Using the OPENMP thread interface

Here it takes the OPENMP path by force, next you see the connected components are all different for the same dataset.

Can you guys check this please, please fix the cmake script to turn off openMP when someone wants.

Regards

**Comment by pmoulon** [comment]
Thank you for your feedback.

1. the UPDATE_CACHE_VARIABLE macro must be defined. See [here](https://github.com/openMVG/openMVG/blob/master/src/cmakeFindModules/UpdateCacheVariable.cmake)

2. USE_OPENMP is never defined or used inside OpenMVG code. The only variable you must use must be OpenMVG_USE_OPENMP.
If you search for USE_OPENMP you will see that this variable is only used in FLANN library https://github.com/openMVG/openMVG/blob/944254c65691b3b7c5fe54d5bdbe312c90de252f/src/third_party/flann/CMakeLists.txt#L57

3. If you want to simulate only one thread on your ubuntu computer, you can easily do it by setting the `export OMP_NUM_THREADS=1` before running openMVG binaries.

4. Which dataset are you using?
I have processed a lot of dataset in windows, linux, mac and always obtained similar results.

I will have a look about the way to put OpenMP to OFF from the cmake command line.

**Comment by fahadzafar** [comment]
Hi

Sorry about not posting the dataset. I was going to make a detailed post, with the dataset and the output ply files once i get home.

Great job on the api though. It works for a lot of things, with the final mesh generation done through MVE.

Regards,
Fahad

**Comment by fahadzafar** [comment]
Hi, I tried to make it as easy as possible to review this, please let me know if I can do something more to help.

- This is the test dataset.
https://drive.google.com/open?id=0BzkvMWM-w80JcEpSdmNYNkJNcUU


- This is the console output when using Ubuntu + OpenMP
https://codepaste.net/7wo8in



- This is the console output when using Mac OS X , without OpenMP. THIS IS WHAT I WANT, AND ITS CORRECT
https://codepaste.net/8zcng4



- This is where the mismatch starts:
https://drive.google.com/open?id=0BzkvMWM-w80Jelp5Zms4ZFEtaUk
The connected components are different, WHY ???????



// ------------------------------------------------------------------
// ------------ MORE INFORMATION 

- This is the cmake config output for Mac OS X
https://codepaste.net/184doa


summary:

** OpenMVG version: 1.1.0
** Build Shared libs: OFF
** Build OpenMVG tests: ON
** Build OpenMVG documentation: ON
** Build OpenMVG samples applications: ON
** Build OpenMVG openGL examples: OFF
** Enable code coverage generation: OFF
** Enable OpenMP parallelization: ON
** Build OpenCV+OpenMVG samples programs: OFF
** Use OpenCV SIFT features: OFF


-- EIGEN: 3.2.8 (internal)
-- CERES: 1.11.0 (internal)
-- FLANN: 1.8.4 (internal)
-- LIBTIFF: 4.0.6 (external)
-- LIBPNG: 1.2.54 (external)
-- LIBJPEG (external)
-- CLP: 1.15.11 (internal)
-- COINUTILS: 2.9.3 (internal)
-- OSI: 0.106.10 (internal)
-- LEMON: 1.3 (internal)


// ------------
// ------------

- CMake config output for Ubuntu.
https://codepaste.net/djg8be

summary:
** OpenMVG version: 1.1.0
** Build Shared libs: OFF
** Build OpenMVG tests: ON
** Build OpenMVG documentation: ON
** Build OpenMVG samples applications: ON
** Build OpenMVG openGL examples: OFF
** Enable code coverage generation: OFF
** Enable OpenMP parallelization: ON
** Build OpenCV+OpenMVG samples programs: OFF
** Use OpenCV SIFT features: OFF


-- EIGEN: 3.2.8 (internal)
-- CERES: 1.11.0 (internal)
-- FLANN: 1.8.4 (internal)
-- LIBTIFF:  (internal)
-- LIBPNG:  (internal)
-- LIBJPEG (internal)
-- CLP: 1.15.11 (internal)
-- COINUTILS: 2.9.3 (internal)
-- OSI: 0.106.10 (internal)
-- LEMON: 1.3 (internal)

**Comment by fahadzafar** [comment]
UPDATE ::: YAY, HAPPY TIMES ....

So thanks to your help, I was able to compile with OpenMVG_USE_OPENMP = OFF. That fixed the problem, using "export OMP_NUM_THREAD=1" didn't run correctly either.

Now, I have confirmed the "fact" that there is a bug when using OpenMP.

Without using OpenMP, you get the same results on Ubuntu and Mac.

Regards,

**Comment by pmoulon** [comment]
Hi @fahadzafar 
OpenMVG_USE_OPENMP = OFF and "export OMP_NUM_THREAD=1" should have been report the same result.

I will check with your data to see what is going on in the coming days.

Did you try with the master branch or the develop branch?

**Comment by pmoulon** [comment]
Hi @fahadzafar 

Here some notes:

1. Using the Sequential pipeline all the cameras are localized:
```
-------------------------------
-- Structure from Motion (statistics):
-- #Camera calibrated: 13 from 13 input images.
-- #Tracks, #3D points: 1490
-------------------------------
SequentialSfMReconstructionEngine::ComputeResidualsMSE.
	-- #Tracks:	1490
	-- Residual min:	1.43683e-05
	-- Residual median:	0.249127
	-- Residual max:	 3.67015
	-- Residual mean:	 0.416025
```

2. You link https://codepaste.net/8zcng4 is not working.

3. Here the result of my test:

Here the output on my Linux (OpenMP on)

```
-----------------------------------------------------------
Global Structure from Motion:
-----------------------------------------------------------
Open Source implementation of the paper:
"Global Fusion of Relative Motions for Robust, Accurate and Scalable Structure from Motion."
Pierre Moulon, Pascal Monasse and Renaud Marlet.  ICCV 2013.
------------------------------------------------------------

- Features Loading -
0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************************************

CleanGraph_KeepLargestBiEdge_Nodes():: => connected Component: 4
Connected component of size: 3
Connected component of size: 1
Connected component of size: 1
Connected component of size: 8
...
Bundle Adjustment statistics (approximated RMSE):
 #views: 13
 #poses: 8
 #intrinsics: 1
 #tracks: 570
 #residuals: 4258
 Initial RMSE: 0.489629
 Final RMSE: 0.453071
 Time (s): 0.0763611

```

Here the output on my Linux (OpenMP set to off in the cmake)

```
 - PUTATIVE MATCHES - 
Use: exhaustive pairwise matching
Using FAST_CASCADE_HASHING_L2 matcher

-----------------------------------------------------------
Global Structure from Motion:
-----------------------------------------------------------
Open Source implementation of the paper:
"Global Fusion of Relative Motions for Robust, Accurate and Scalable Structure from Motion."
Pierre Moulon, Pascal Monasse and Renaud Marlet.  ICCV 2013.
------------------------------------------------------------

- Features Loading -
0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************************************

CleanGraph_KeepLargestBiEdge_Nodes():: => connected Component: 5
Connected component of size: 1
Connected component of size: 1
Connected component of size: 1
Connected component of size: 1
Connected component of size: 9

...

Bundle Adjustment statistics (approximated RMSE):
 #views: 13
 #poses: 9
 #intrinsics: 1
 #tracks: 573
 #residuals: 4378
 Initial RMSE: 0.514768
 Final RMSE: 0.486775
 Time (s): 0.0904023


```

So you are right the results are a bit different.
I need to investigate to understand if the difference is:
- at matching stage (putatives matches)
- or at putative matches filtering stage (it could be due to RANSAC results that are based randomness)
- or due to multithreading (task ordering issue).

5. If you want more accuracy at the matching stage you can use the ANNL2 matcher
```
print ("3. Compute matches")
pMatches = subprocess.Popen( [os.path.join(OPENMVG_SFM_BIN, "openMVG_main_ComputeMatches"),  "-i", matches_dir+"/sfm_data.json", "-o", matches_dir, "-g", "e", "-n", "ANNL2"] )
```

It will be a bit slower but the accuracy will be better

**Comment by rperrot** [comment]
Is the difference related to all matching algorithms (Fast cascade, cascade, ann ?) or only to a specific one ?

**Comment by fahadzafar** [comment]
Hi @pmoulon , first off thanks for helping out with this and doing it so quickly,

- I corrected the link, so the correct run on Mac OS X output is viewable.

- You presented the output with OPENMP on, but during the putative matches stage, i don't see "Using OPENMP thread interface" on your console output, for instance this is my run with openmp,

` - PUTATIVE MATCHES - 
Use: exhaustive pairwise matching
Using FAST_CASCADE_HASHING_L2 matcher
Using the OPENMP thread interface`

- There is a difference in the output with and without OpenMP, and subtle differences are ok, We all know processors multiple doubles and floats differently, but here is my actual issue. The difference is so large, that in the next stage of my pipeline, OpenMVE discards the point cloud as not good enough to get a mesh out, while the good run on the same dataset  gives me a mesh through OpenMVE, and "texrecon" is able to create a textured mesh. I have some instances where, when OMP=ON, i get a mesh but "texrecon" will exit out throwing some kind of error. This never happens with the OMP=OFF execution unless the images have less correlation. This I thought was alarming.

**Comment by rperrot** [comment]
"There is a difference in the output with and without OpenMP, and subtle differences are ok" 

In the way multithreading is processed in openMVG (it's parallel job processing only, theoretically independent of the processing order), there must be a 1:1 relation in the output of the multithreading and in the output of the mono threading. You pointed out a but not a desirable behavior. 

You only tested the FAST_CASCADE_HASHING or any other matching method ?

**Comment by fahadzafar** [comment]
I only tried FAST_CASCADE_HASHING, 

- I agree that multi-threading should have the same results, but unless there is a test that checks it, there are always mismatches. I have seen code where there are minor overflows when people use threading,. Specially with OMP, the pragmas can cause havoc if you are using [i-1] or [i+1] values in the i(th) iteration. Its easy to miss since OMP parallelization is so easy to insert.


- using the MainBranch.

**Comment by rperrot** [comment]
"I am working with the main github branch."

That's an important information ! Thanks. 

"I have seen code where there are minor overflows when people"

You're right in case of summing elements in an array. (summing without EFT is prone to imprecision, canceling, rounding, underflow and over fun with floating point). I repeat, in openMVG case, matching is done in parallel at the level of the images (i.e. we compute matching of image 1-2 ; 1-3 ; 1-4 in parallel using the same algorithm, matching between image pair), so the result should be the same in any order and in any SMP configuration.

**Comment by fahadzafar** [comment]
oh I get it. 
Simply put, there should be no ghost region issues if parallelization is not patch-based but rather image based. Then it seems like a very minor issue somewhere.

**Comment by rperrot** [comment]
Then it seems like a very minor issue somewhere.

Yes, maybe a single pragma critical that is missing when inserting result of matching into the global matching container.

**Comment by naveenjafer** [comment]
I am facing Similar issues with my datasets. To understand the issue, I have taken a dataset with only 2 images. I ran incremental SFM on them once with single thread and once with multiple threads. I almost always attain a successful reconstruction with single thread
`Track stats
------------------
-- Tracks Stats --
 Tracks number: 2866
 Images Id: 
0, 1, 
------------------
TrackLength, Occurrence
	2	2866

Automatic selection of an initial pair:
0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************************************
A-Contrario initial pair residual: 5.62567

Bundle Adjustment statistics (approximated RMSE):
 #views: 2
 #poses: 2
 #intrinsics: 1
 #tracks: 2866
 #residuals: 11464
 Initial RMSE: 0.228648
 Final RMSE: 0.218397
 Time (s): 0.0441466




SequentialSfMReconstructionEngine::ComputeResidualsMSE.
	-- #Tracks:	2864
	-- Residual min:	2.44345e-08
	-- Residual median:	0.0176139
	-- Residual max:	 2.69691
	-- Residual mean:	 0.175167

=========================
 MSE Residual InitialPair Inlier: 0.175167
=========================


-------------------------------
-- Structure from Motion (statistics):
-- #Camera calibrated: 2 from 2 input images.
-- #Tracks, #3D points: 2864
-------------------------------



SequentialSfMReconstructionEngine::ComputeResidualsMSE.
	-- #Tracks:	2864
	-- Residual min:	2.44345e-08
	-- Residual median:	0.0176139
	-- Residual max:	 2.69691
	-- Residual mean:	 0.175167

Histogram of residuals:

0	|	8350
0.27	|	2029
0.539	|	718
0.809	|	254
1.08	|	89
1.35	|	8
1.62	|	4
1.89	|	0
2.16	|	0
2.43	|	3
2.7


 Total Ac-Sfm took (s): 2.048
...Generating SfM_Report.html
...Export SfM_Data to disk.

`

With the multi threaded one though, this is the output I get.
`Track stats
------------------
-- Tracks Stats --
 Tracks number: 10773
 Images Id: 
0, 1, 
------------------
TrackLength, Occurrence
	2	10773

Automatic selection of an initial pair:
0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************************************

----------------------------------------------------
SequentialSfMReconstructionEngine::ChooseInitialPair
----------------------------------------------------
 Pairs that have valid intrinsic and high support of points are displayed:
 Choose one pair manually by typing the two integer indexes
----------------------------------------------------

(0,1)		10773 matches

Putative starting pair is: (0,1)
A-Contrario initial pair residual: 3.45786e-24

Bundle Adjustment statistics (approximated RMSE):
 #views: 2
 #poses: 2
 #intrinsics: 1
 #tracks: 10773
 #residuals: 43092
 Initial RMSE: 1.88379e-13
 Final RMSE: 1.88379e-13
 Time (s): 0.0531171


=========================
 MSE Residual InitialPair Inlier: -1
=========================

`

The number of matches is considerably different in both cases and the Initial Pair choose engine is fired only in the second case(I had changed the code to automatically select the pair with the highest matches, hence no prompt for input from user is seen). Is this some problem with the matcher under openMP?

**Comment by pmoulon** [comment]
Seems like there is a glitch somewhere.
I will try to investigate what is happening.
@naveenjafer Do you can share to me your two images?

**Comment by naveenjafer** [comment]
@pmoulon images 100_7104.JPG and 100_7015.JPG are two good examples from the SceauxCastle dataset that fail when run with multiple threads and work fine when run with single thread. Parameters given were as follows. SIFT - HIGH - Incremental. In fact almost all the pairs in this dataset and others exhibit similar behaviour under multiple threads.

**Comment by pmoulon** [comment]
Here my observations:

1.

It's really strange that in one case you have `-- Tracks Stats --
Tracks number: 2866` and in the other `-- Tracks Stats --
Tracks number: 10773`. Even by using SIFT ULTRA I don't have this number of tracks.

2.

I tested (develop branch) with and without OpenMP and I obtain the same result. I'm not able to reproduce your behavior.

I used a scene with image 100_7104.JPG and 100_7005.JPG (default settings):

**WithoutOpenMP:**
```
-------------------------------
-- Structure from Motion (statistics):
-- #Camera calibrated: 2 from 2 input images.
-- #Tracks, #3D points: 1279
-------------------------------
```

**With OpenMP**
```
-------------------------------
-- Structure from Motion (statistics):
-- #Camera calibrated: 2 from 2 input images.
-- #Tracks, #3D points: 1279
-------------------------------
```

I used a scene with image 100_7104.JPG and 100_7005.JPG (settings with SIFT - HIGH):

**(no openMP)**
```
-------------------------------
-- Structure from Motion (statistics):
-- #Camera calibrated: 2 from 2 input images.
-- #Tracks, #3D points: 2826
-------------------------------
```

**(with openMP)**
```
-------------------------------
-- Structure from Motion (statistics):
-- #Camera calibrated: 2 from 2 input images.
-- #Tracks, #3D points: 2882
-------------------------------
```

3.
Can you test again on the develop branch on your side too?

**Comment by pmoulon** [comment]
@fahadzafar Please can you try with the develop branch?

**Comment by naveenjafer** [comment]
@pmoulon I'll try it out with the dev branch and get back to you soon.

**Comment by fahadzafar** [comment]
Sure, will update you on this using the develop branch.

**Comment by naveenjafer** [comment]
@pmoulon I seem to have no issues making the master branch, but trying to make the develop branch is giving me errors.

```
/home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc: In member function ‘void image_undistort_gui::MainWindow::BuildMenus()’:
/home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:709:36: error: no matching function for call to ‘QAction::QAction(const char [5])’
   m_quit_act = new QAction( "Quit" );
                                    ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtWidgets/QAction:1:0,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.hh:15,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:8:
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note: candidate: QAction::QAction(QActionPrivate&, QObject*)
     QAction(QActionPrivate &dd, QObject *parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note: candidate: QAction::QAction(const QIcon&, const QString&, QObject*)
     QAction(const QIcon &icon, const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note:   candidate expects 3 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note: candidate: QAction::QAction(const QString&, QObject*)
     QAction(const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note: candidate: QAction::QAction(QObject*)
     explicit QAction(QObject* parent);
              ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note:   no known conversion for argument 1 from ‘const char [5]’ to ‘QObject*’
/home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:714:48: error: no matching function for call to ‘QAction::QAction(const char [11])’
   m_open_image_act = new QAction( "Open image" );
                                                ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtWidgets/QAction:1:0,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.hh:15,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:8:
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note: candidate: QAction::QAction(QActionPrivate&, QObject*)
     QAction(QActionPrivate &dd, QObject *parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note: candidate: QAction::QAction(const QIcon&, const QString&, QObject*)
     QAction(const QIcon &icon, const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note:   candidate expects 3 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note: candidate: QAction::QAction(const QString&, QObject*)
     QAction(const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note: candidate: QAction::QAction(QObject*)
     explicit QAction(QObject* parent);
              ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note:   no known conversion for argument 1 from ‘const char [11]’ to ‘QObject*’
/home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:716:44: error: no matching function for call to ‘QAction::QAction(const char [9])’
   m_use_grid_act = new QAction( "Use grid" );
                                            ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtWidgets/QAction:1:0,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.hh:15,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:8:
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note: candidate: QAction::QAction(QActionPrivate&, QObject*)
     QAction(QActionPrivate &dd, QObject *parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note: candidate: QAction::QAction(const QIcon&, const QString&, QObject*)
     QAction(const QIcon &icon, const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note:   candidate expects 3 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note: candidate: QAction::QAction(const QString&, QObject*)
     QAction(const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note: candidate: QAction::QAction(QObject*)
     explicit QAction(QObject* parent);
              ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note:   no known conversion for argument 1 from ‘const char [9]’ to ‘QObject*’
/home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:721:50: error: no matching function for call to ‘QAction::QAction(const char [14])’
   m_no_dist_act   = new QAction( "No distortion" );
                                                  ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtWidgets/QAction:1:0,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.hh:15,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:8:
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note: candidate: QAction::QAction(QActionPrivate&, QObject*)
     QAction(QActionPrivate &dd, QObject *parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note: candidate: QAction::QAction(const QIcon&, const QString&, QObject*)
     QAction(const QIcon &icon, const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note:   candidate expects 3 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note: candidate: QAction::QAction(const QString&, QObject*)
     QAction(const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note: candidate: QAction::QAction(QObject*)
     explicit QAction(QObject* parent);
              ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note:   no known conversion for argument 1 from ‘const char [14]’ to ‘QObject*’
/home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:723:45: error: no matching function for call to ‘QAction::QAction(const char [9])’
   m_radial_k1_act = new QAction( "Radial 1" );
                                             ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtWidgets/QAction:1:0,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.hh:15,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:8:
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note: candidate: QAction::QAction(QActionPrivate&, QObject*)
     QAction(QActionPrivate &dd, QObject *parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note: candidate: QAction::QAction(const QIcon&, const QString&, QObject*)
     QAction(const QIcon &icon, const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note:   candidate expects 3 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note: candidate: QAction::QAction(const QString&, QObject*)
     QAction(const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note: candidate: QAction::QAction(QObject*)
     explicit QAction(QObject* parent);
              ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note:   no known conversion for argument 1 from ‘const char [9]’ to ‘QObject*’
/home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:725:45: error: no matching function for call to ‘QAction::QAction(const char [9])’
   m_radial_k3_act = new QAction( "Radial 3" );
                                             ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtWidgets/QAction:1:0,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.hh:15,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:8:
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note: candidate: QAction::QAction(QActionPrivate&, QObject*)
     QAction(QActionPrivate &dd, QObject *parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note: candidate: QAction::QAction(const QIcon&, const QString&, QObject*)
     QAction(const QIcon &icon, const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note:   candidate expects 3 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note: candidate: QAction::QAction(const QString&, QObject*)
     QAction(const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note: candidate: QAction::QAction(QObject*)
     explicit QAction(QObject* parent);
              ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note:   no known conversion for argument 1 from ‘const char [9]’ to ‘QObject*’
/home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:727:46: error: no matching function for call to ‘QAction::QAction(const char [14])’
   m_brown_act = new QAction( "Brown (3R-2T)" );
                                              ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtWidgets/QAction:1:0,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.hh:15,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:8:
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note: candidate: QAction::QAction(QActionPrivate&, QObject*)
     QAction(QActionPrivate &dd, QObject *parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note: candidate: QAction::QAction(const QIcon&, const QString&, QObject*)
     QAction(const QIcon &icon, const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note:   candidate expects 3 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note: candidate: QAction::QAction(const QString&, QObject*)
     QAction(const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note: candidate: QAction::QAction(QObject*)
     explicit QAction(QObject* parent);
              ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note:   no known conversion for argument 1 from ‘const char [14]’ to ‘QObject*’
/home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:729:42: error: no matching function for call to ‘QAction::QAction(const char [8])’
   m_fisheye_act = new QAction( "Fisheye" );
                                          ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtWidgets/QAction:1:0,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.hh:15,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:8:
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note: candidate: QAction::QAction(QActionPrivate&, QObject*)
     QAction(QActionPrivate &dd, QObject *parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note: candidate: QAction::QAction(const QIcon&, const QString&, QObject*)
     QAction(const QIcon &icon, const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note:   candidate expects 3 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note: candidate: QAction::QAction(const QString&, QObject*)
     QAction(const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note: candidate: QAction::QAction(QObject*)
     explicit QAction(QObject* parent);
              ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note:   no known conversion for argument 1 from ‘const char [8]’ to ‘QObject*’
/home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:738:44: error: no matching function for call to ‘QAction::QAction(const char [10])’
   m_process_act = new QAction( "Undistort" );
                                            ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtWidgets/QAction:1:0,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.hh:15,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:8:
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note: candidate: QAction::QAction(QActionPrivate&, QObject*)
     QAction(QActionPrivate &dd, QObject *parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note: candidate: QAction::QAction(const QIcon&, const QString&, QObject*)
     QAction(const QIcon &icon, const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note:   candidate expects 3 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note: candidate: QAction::QAction(const QString&, QObject*)
     QAction(const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note: candidate: QAction::QAction(QObject*)
     explicit QAction(QObject* parent);
              ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note:   no known conversion for argument 1 from ‘const char [10]’ to ‘QObject*’
/home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:739:47: error: no matching function for call to ‘QAction::QAction(const char [13])’
   m_export_act  = new QAction( "Export image" );
                                               ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtWidgets/QAction:1:0,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.hh:15,
                 from /home/rajas/Desktop/2NaveenOpenMVGDev/openMVG/src/openMVG_Samples/image_undistort_gui/MainWindow.cc:8:
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note: candidate: QAction::QAction(QActionPrivate&, QObject*)
     QAction(QActionPrivate &dd, QObject *parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:174:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note: candidate: QAction::QAction(const QIcon&, const QString&, QObject*)
     QAction(const QIcon &icon, const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:89:5: note:   candidate expects 3 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note: candidate: QAction::QAction(const QString&, QObject*)
     QAction(const QString &text, QObject* parent);
     ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:88:5: note:   candidate expects 2 arguments, 1 provided
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note: candidate: QAction::QAction(QObject*)
     explicit QAction(QObject* parent);
              ^
/usr/include/x86_64-linux-gnu/qt5/QtWidgets/qaction.h:87:14: note:   no known conversion for argument 1 from ‘const char [13]’ to ‘QObject*’
openMVG_Samples/image_undistort_gui/CMakeFiles/openMVG_sample_image_undistort.dir/build.make:86: recipe for target 'openMVG_Samples/image_undistort_gui/CMakeFiles/openMVG_sample_image_undistort.dir/MainWindow.cc.o' failed
make[2]: *** [openMVG_Samples/image_undistort_gui/CMakeFiles/openMVG_sample_image_undistort.dir/MainWindow.cc.o] Error 1
CMakeFiles/Makefile2:3466: recipe for target 'openMVG_Samples/image_undistort_gui/CMakeFiles/openMVG_sample_image_undistort.dir/all' failed
make[1]: *** [openMVG_Samples/image_undistort_gui/CMakeFiles/openMVG_sample_image_undistort.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by pmoulon** [comment]
QT > 5.4 is required...

Else you can comment the compilation of this binary here:
https://github.com/openMVG/openMVG/blob/develop/src/openMVG_Samples/CMakeLists.txt#L24

- [ ] I need to add a OpenMVG_BUILD_GUI cmake variable in order to activate or disable the QT GUI based build

**Comment by zwn** [comment]
@pmoulon Could the existence and version of QT be detected by cmake and only build the targets if QT>5.4 is found?

**Comment by rperrot** [comment]
@zwn 

Test for presence of Qt is already done : https://github.com/openMVG/openMVG/blob/develop/src/openMVG_Samples/image_undistort_gui/CMakeLists.txt

The issue is here with the version of Qt to use. (we also need to test Qt5Widgets_VERSION variable).

**Comment by zwn** [comment]
There are more of these. See for example 
https://github.com/openMVG/openMVG/blob/develop/src/software/CMakeLists.txt#L13
https://github.com/openMVG/openMVG/blob/develop/src/openMVG_Samples/describe_and_match_GUI/CMakeLists.txt#L3
cmake is complaining about QT every time I run it (finding v4 while looking for v5), I was just ignoring it, trying to focus on other things.

**Comment by naveenjafer** [comment]
@pmoulon Thanks for that, yes you are right, this issue is not happening with the developer version, It is reproducable only with the master version. Kindly look into the same if you have time.

**Comment by MatillaMartin** [comment]
@pmoulon I came across a similar issue, different results with multithreading on and off (using -n in main_ComputeFeatures). It seems that the features are not computed correctly for some images, and are mixed bewteen images. I will post some files soon, but its a quick tip so you know where to start looking

**Comment by pmoulon** [comment]
Reproducibility have been enhanced in the develop branch.
